src/Autofac/Core/Activators/ProvidedInstance/ProvidedInstanceActivator.cs

```diff
@@ -64,23 +67,62 @@ private object GetInstance()
         /// </summary>
         public bool DisposeInstance { get; set; }
 
-        /// <summary>
-        /// Releases unmanaged and - optionally - managed resources.
-        /// </summary>
-        /// <param name="disposing"><c>true</c> to release both managed and unmanaged resources; <c>false</c> to release only unmanaged resources.</param>
+        /// <inheritdoc />
         protected override void Dispose(bool disposing)
         {
             // Only dispose of the instance here if it wasn't activated. If it was activated,
             // then either the owning lifetime scope will dispose of it automatically
             // (see InstanceLookup.Activate) or an OnRelease handler will take care of it.
-            if (disposing && DisposeInstance && _instance is IDisposable disposable && !_activated)
+            if (disposing && DisposeInstance && !_activated)
             {
-                disposable.Dispose();
+                // If we are in synchronous dispose, and an object implements IDisposable,
+                // then use it.
+                if (_instance is IDisposable disposable)
+                {
+                    disposable.Dispose();
+                }
+                else if (_instance is IAsyncDisposable asyncDisposable)
+                {
+                    Trace.TraceWarning(DisposerResources.TypeOnlyImplementsIAsyncDisposable, LimitType.FullName);
+
+                    // Type only implements IAsyncDisposable. We will need to do sync-over-async.
+                    // We want to ensure we lose all context here, because if we don't we can deadlock.
+                    // So we push this disposal onto the threadpool.
+                    Task.Run(async () => await asyncDisposable.DisposeAsync().ConfigureAwait(false))
+                        .ConfigureAwait(false)
+                        .GetAwaiter().GetResult();
+                }
             }
 
             base.Dispose(disposing);
         }
 
+        /// <inheritdoc />
+        protected override async ValueTask DisposeAsync(bool disposing)
+        {
+            if (disposing && DisposeInstance && !_activated)
+            {
+                // If the item implements IAsyncDisposable we will call its DisposeAsync Method.
+                if (_instance is IAsyncDisposable asyncDisposable)
+                {
+                    var vt = asyncDisposable.DisposeAsync();
+
+                    // Don't await if it's already completed (this is a slight gain in performance of using ValueTask).
```

Comment:
there is no perf gain as generated state machine is doing exactly the same. It's worth doing that manually only in non-async functions (this avoids creating state machine).

Reply:
@OwnageIsMagic While we appreciate folks getting more eyes on the code, I'm not sure what the goal is with some of the comments here. If you have a pull request you'd like to submit to help improve the product, awesome, we'd love to check it out; however, many of these things are months or years completed, so we're not likely to go back and take any action on them at this time.

Reply:
I'm just backporting stuff to netfx and looking through some commits.
Before commenting I check develop branch and the code in question is still there, so it quite actionable.
For now I'm just informing you that this comment here is plain wrong, so you (and other maintainers) can consider not doing this in the future (sharing knowledge).
Regardless my [other comment](https://github.com/autofac/Autofac/pull/1276#issuecomment-1243804201), I think it's have more sense it context of old issue, rather then creating new issue for it.

Reply:
@tillig as I said it's not _something from a commit months or years ago_.
It right there, in the HEAD of develop (no pun intended) https://github.com/autofac/Autofac/blob/develop/src/Autofac/Core/Activators/ProvidedInstance/ProvidedInstanceActivator.cs#L106
I guess that thing you are trying to say, is that this style of contribution doesn't fit into your workflow and I should open a new issue.

Reply:
It may be in `develop` but scroll up and you'll see this PR with this commit is from August 2021, so, yes, it is from over a year ago. But, also, correct, comments on code without any context or issue filed are not as helpful in our flow as it may be in other repositories. There isn't a big team here like on ASP.NET Core or the .NET Runtime and no one here's getting paid. Given that, if folks want action, we have to be a little more specific about how the intake process works. It's great to get more eyes on the code, but if something needs to change or if there's a bug, something incorrect, whatever, it really needs to be an issue so we can prioritize it. Comments like this are potentially interesting discussion points, but likely won't result in anything but discussion, and, as noted, we don't have a ton of bandwidth given the small size of the team, so it may not even result in discussion.